openide.awt/test/unit/src/org/netbeans/modules/openide/awt/StatefulActionProcessorTest.java

```diff
@@ -0,0 +1,1386 @@
+        mod = null;
+        assertGC("Action model must be GCed", r);
+    }
+    
+    /**
+     * Check that action with action check enables only after the actual instance enables.
+     */
+    public void testCustomEnableActionInstantiation() {
+        assertEquals("Not pre-created", 0, created);
+        Action a = Actions.forID("Foo", "test.CustomEnableAction");
+        assertNotNull(a);
+        assertEquals("Not direcly created from layer", 0, created);
+        a.isEnabled();
+        assertEquals("Not created unless data present", 0, created);
+        
+        ActionModel3 mod = new ActionModel3();
+        lookupContent.add(mod);
+        
+        // still not enabled
+        assertFalse(a.isEnabled());
+        assertEquals("Not created unless guard is set", 0, created);
+        
+        mod.boolProp = true;
+        mod.supp.firePropertyChange("boolProp", null, null);
+        // should instantiate the action just because of the property change on guard,
+        // now the action decides the final state.
+        assertEquals("Must be created to evaluate enabled state", 1, created);
+        
+        Action inst = instance;
+        assertNotNull(inst);
+        
+        inst.setEnabled(true);
+        assertSame("Same instance for repeated enable", inst, instance);
+        
+        a.actionPerformed(new ActionEvent(this, 0, "cmd"));
+        assertSame("Same instance for invocation and enable eval", inst, instance);
+        assertNotNull(received);
+        assertEquals("cmd", received.getActionCommand());
+    }
+    
+    /**
+     * Checks that when the context object is changed, the old custom
+     * action instance is trashed and a new one is created.
+     */
+    public void testCustomEnableActionChangesWithContext() {
+        Action a = Actions.forID("Foo", "test.CustomEnableAction");
+        ActionModel3 mod = new ActionModel3();
+        mod.boolProp = true;
+
+        lookupContent.add(mod);
+        CustomEnableAction inst = (CustomEnableAction)instance;
+        assertNotNull(inst);
+        assertSame(mod, inst.model);
+        assertFalse(a.isEnabled());
+        
+        inst.setEnabled(true);
+        assertTrue(a.isEnabled());
+        
+        ActionModel3 mod2 = new ActionModel3();
+        instance = null;
+        lookupContent.remove(mod);
+        lookupContent.add(mod2);
+        assertNull(instance);
+
+        mod2.boolProp = true;
+        mod2.supp.firePropertyChange("boolProp", null, null);
+        
+        assertNotNull(instance);
+        Action save2 = instance;
+        
+        assertFalse(a.isEnabled());
+        ((AbstractAction)instance).setEnabled(true);
+        a.actionPerformed(new ActionEvent(this, 0, "x"));
+        
+        assertSame(instance, save2);
+        assertNotSame(instance, inst);
+    }
+    
+    @ActionID(id = "test.ToggleAction", category="Foo")
+    @ActionRegistration(displayName = "TestAction")
+    @ActionState(property = "boolProp")
```

Comment:
@sdedic, is there any value of declaring the `@ActionState` as a separate annotation? Wouldn't it be more more consistent to have:
```
    @ActionID(id = "test.ToggleAction", category="Foo")
    @ActionRegistration(displayName = "TestAction",
       checkOn=@ActionState(property = "boolProp")
    )
```
I mean there is an asymetry between enabled and checked state definitions. It is shifted in favor of checked state and I am not sure it is correct, as I expect even more usages of the `@ActionState` to describe the enabled state.

Reply:
There's no reason for annotation directly on the element; except the evolution of the prototype code :) I have started with "checked" state, then realized the almost same mechanism can be applied to enabled state.

Will change to `@ActionRegistration`'s value.